### PR TITLE
fix(utils): add is binary icon

### LINKED ISSUE #592: Show `isBinary` icon report formats and stdout
### User story

I configured a category as binary with the flag `isBinay` in `CategoryConfig`.

To test is I ran the collect command.

Screenshot 2024-03-29 at 23 07 36

### Acceptance criteria

- [ ] If a category is binary I would expect to get a visual feedback or influence on the rendered score.
  - [ ] In the **terminal** output in the **category overview** we show the checkmark/cross **before** the score number.  
   e.g. `✗ 87` or `✓ 87`.
  - [ ] In the **markdow report** in the **category overview** section we show the checkmark/cross **before** the score number.  
   e.g. `🟡 **87** ❌` or `🟡 **87** ✅`.
  - [ ] In the **markdow report** in the **category detail** section we show the checkmark/cross **after** the score number.  
 e.g. `Score:  **87** ❌` or `Score:  **87** ✅`.
   - [ ] In the **comparison report** in the **category overview** section we show the checkmark/cross **after** the score.  
 e.g. `🟡 **87** ❌` or `🟡 **87** ✅`.

### Implementation details

I suggest to already consider the new `targetScore` use naming that targets this future feature.

**`targetScore` icon logic suggestion**

```ts
export function targetScoreIcon(score: number, targetScore?: number): string {
  if (targetScore != null) {
    if (score >= targetScore) {
      return `✅`;
    }
    return `❌`;
  }
  return '';
}

```

---

## Examples

**Terminal Table**

```
┌─────────────────────────────────────────────────────────┬──────────┬──────────┐
│  Category                                               │  Score   │  Audits  │
├─────────────────────────────────────────────────────────┼──────────┼──────────┤
│  Bug prevention                                         │   ✓ 100  │     108  │
├─────────────────────────────────────────────────────────┼──────────┼──────────┤
│  Code style                                             │    ✗ 99  │     231  │
└─────────────────────────────────────────────────────────┴──────────┴──────────┘

```

Screenshot 2024-06-23 at 21 16 06

---

**Report: Category Overview**

|🏷 Category|⭐ Score|🛡 Audits|
|:--|:--:|:--:|
| [Bug prevention](#bug-prevention) |🟢 **100** ✅|56|
| [Code style](#code-style) |🟢 **99** ❌|73|
| [Bug prevention](#bug-prevention) |🟢 **100 ✓**|56|
| [Code style](#code-style) |🟢 **99 ✗**|73|

---  

**Report: Category Details**

### Bug prevention

🟢 Score:  **100** ✅
- 🟩 [no-any](#no-any) (_eslint_) - **0**

### Code style

🟢 Score:  **99** ❌
- 🟩 [no-let) (_eslint_) - **0**

### Bug prevention

🟢 Score:  **100 ✓** 
- 🟩 [no-any](#no-any) (_eslint_) - **0**

### Code style

🟢 Score:  **99 ✗**
- 🟩 [no-let) (_eslint_) - **0**

---

**Comparison comment suggestion**

# Code PushUp

🤨 Code PushUp report has both **improvements and regressions** – compared target commit e214a030d91484a29071850ba811f5b43f1a3c16 with source commit 8d6dccbdf4799e97ce83b8909bacf96fc7ad8de8.

## 🏷️ Categories

|🏷️ Category|⭐ Current score|⭐ Previous score|🔄 Score change|
|:--|:--:|:--:|:--:|
| Bug prevention |🟡 **99** ✅|🟡 100 ❌||
| Code style |🟡 **98** ✅|🟡 99 ❌||
| Bug prevention |🟡 **100 ✓** |🟡 99 ✗||
| Code style |🟡 **98 ✗** |🟡 98 ✗||

</details>

### PR DESCRIPTION
**Pull request by BioPhoton**
closes #592 

Suggested followup #767

### CODE REVIEW

File: packages/utils/src/lib/reports/generate-md-report-categoy-section.ts [hunk 1]
```diff
@@ -23,10 +23,15 @@ export function categoriesOverviewSection(
...
+    categories.map(({ title, refs, score, isBinary }) => [
+      // @TODO refactor `isBinary: boolean` to `targetScore: number` #713
       // The heading "ID" is inferred from the heading text in Markdown.
       md.link(`#${slugify(title)}`, title),
-      md`${scoreMarker(score)} ${md.bold(formatReportScore(score))}`,
+      md`${scoreMarker(score)} ${md.bold(
+        formatReportScore(score),
+      )}${targetScoreIcon(score, isBinary === true ? 1 : undefined, {
+        prefix: ' ',
+      })}`,
```
Comment (matejchalk) [inline packages/utils/src/lib/reports/generate-md-report-categoy-section.ts:34]: These calls could be made more compact by introducing a simple wrapper function:

```ts
export function binaryIconSuffix(score: number, isBinary: boolean | undefined): string {
  return targetScoreIcon(score, isBinary ? 1 : undefined, { prefix: ' ' });
}
```

```suggestion
      )}${binaryIconSuffix(score, isBinary)}`,
```

Also applies to lines 54-57 below.

File: packages/utils/src/lib/reports/log-stdout-summary.ts [hunk 2]
```diff
@@ -65,11 +64,19 @@ function logPlugins(report: ScoredReport): void {
...
+
+  const rows = categories.map(({ title, score, refs, isBinary }) => [
     title,
-    applyScoreColor({ score }),
+    // @TODO refactor `isBinary: boolean` to `targetScore: number` #713
+    `${targetScoreIcon(score, isBinary === true ? 1 : undefined, {
+      passIcon,
+      failIcon,
+      postfix: ' ',
+    })}${applyScoreColor({ score })}`,
```
Comment (matejchalk) [inline packages/utils/src/lib/reports/log-stdout-summary.ts:79]: Similar here, multiline function calls in template strings are quite hard to read, so I would extract it to a helper for the stdout reports. Something like this:

```ts
function binaryIconPrefix(score: number, isBinary: boolean | undefined): string {
  return targetScoreIcon(score, isBinary ? 1 : undefined, {
    passIcon: bold(green('✓')),
    failIcon: bold(red('✗')),
    postfix: ' ',
  });
}

```

```suggestion
    `${binaryIconPrefix(score, isBinary)}${applyScoreColor({ score })}`,
```

File: packages/utils/src/lib/reports/utils.ts [hunk 3]
```diff
@@ -324,3 +253,52 @@ export function compareIssues(a: Issue, b: Issue): number {
...
+    postfix?: string;
+  } = {},
+): string {
+  if (targetScore != null) {
+    const {
+      passIcon = '✅',
+      failIcon = '❌',
+      prefix = '',
+      postfix = '',
+    } = options; // ✅✓ ❌✗
```
Comment (matejchalk) [inline packages/utils/src/lib/reports/utils.ts:297]: Not sure what to make of this comment :smile:

File: packages/utils/src/lib/reports/sorting.ts [hunk 4]
```diff
@@ -9,10 +14,81 @@ import {
...
+          weight: ref.weight,
+          type: 'audit',
+        },
+        plugins,
+      ),
+    )
+    .sort(compareCategoryAuditsAndGroups);
+}
+
+export function getSortableGroupByRef(
```
Comment (matejchalk) [inline packages/utils/src/lib/reports/sorting.ts:61]: Good call on extracting these :+1:

File: packages/utils/src/lib/reports/utils.ts (hunk 3 above)
Comment (BioPhoton) [reply packages/utils/src/lib/reports/utils.ts:297]: removed

File: packages/utils/src/lib/reports/generate-md-report-categoy-section.ts (hunk 1 above)
Comment (BioPhoton) [reply packages/utils/src/lib/reports/generate-md-report-categoy-section.ts:34]: implemented the function and tests.

File: packages/utils/src/lib/reports/__snapshots__/generate-md-report-category-section.unit.test.ts.snap [hunk 5]
```diff
@@ -1,11 +1,15 @@
 // Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html
 
-exports[`categoriesDetails > should render complete categories details 1`] = `
+exports[`categoriesDetailsSection > should render categories details and add "✅" when isBinary is passing 1`] = `_MarkdownDocument {}`;
+
+exports[`categoriesDetailsSection > should render categories details and add "❌" when isBinary is failing 1`] = `_MarkdownDocument {}`;
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply packages/utils/src/lib/reports/__snapshots__/generate-md-report-category-section.unit.test.ts.snap]: fixed

**Review by matejchalk — APPROVED** [review]

**Pull request merged**